File: c++/swapPairs.cpp
#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x) : val(x), next(NULL) {}
};

ListNode* buildList(int a[], int n);

ListNode *swapPairs(ListNode *head) {
    if (!head) return NULL;
    else if (head && !head->next) return head;
    
    ListNode *prev, *curr, *futu;
    prev = head;
    curr = head->next;
    head = curr;
    futu = curr->next;
    curr->next = prev;
    prev = curr;
    curr = curr->next;
    curr->next = futu;
    while(futu && futu->next)  {
        prev = curr;
        curr = curr->next;
        prev->next = futu->next;
        prev = prev->next;
        futu = futu->next->next;
        prev->next = curr;
        curr->next = futu;
    }
    if (!futu) return head;
    if (futu && !futu->next) {
        curr->next = futu;
        return head;
    }
}

int main() {
    int n = 4;
    int a[] = {1, 2, 3, 4};
    ListNode* head = buildList(a, n);
    cout << endl;
    ListNode* curr = head;
    while(curr)  {
        cout << curr->val << ", ";
        curr = curr->next;
    }

    ListNode* result = swapPairs(head);

    cout << endl;
    while(result)  {
        cout << result->val << ", ";
        result = result->next;
    }
    return 0;
}

ListNode* buildList(int a[], int n) {
    ListNode* head, *curr;
    for (int i = 0; i < n; ++i) {
        if (i == 0)  {
            head = new ListNode(a[i]);
            curr = head;
        } else  {
            curr->next = new ListNode(a[i]);
            curr = curr->next;
        }
    }
    return head;
}

File: java/swapPairs.java
import com.ListNode;

import java.util.HashMap;
import java.util.Map;
import java.util.ArrayList;
import java.util.List;

public class swapPairs {
    public static class Solution {
        public ListNode swapPairs0(ListNode head) {
            if (head == null || head.next == null) return head;
            ListNode result = head.next;
            ListNode fast = result.next;
            ListNode curr = head;
            result.next = curr;
            curr.next = fast;
            ListNode prev = result;
            while (fast != null) {
                if (fast.next == null)
                    return result;
                fast = fast.next.next;
                prev = curr;
                curr = curr.next.next;
                curr.next = prev.next;
                prev.next = curr;
                prev = curr;
                curr = curr.next;
                curr.next = fast;
            }
            return result;
        }

        public ListNode swapPairs1(ListNode head) {
            if (head == null || head.next == null) return head;
            ListNode dummy = new ListNode(-1);
            dummy.next = head;
            for (ListNode prev = dummy, curr = prev.next, next = curr.next;
                 next != null;
                 prev = curr, curr = curr.next, next = curr != null ? curr.next : null) {
                prev.next = next;
                curr.next = next.next;
                next.next = curr;
            }
            return dummy.next;
        }

        public ListNode swapPairs(ListNode head) {
            ListNode p = head;
            int tmp = 0;
            while (p != null && p.next != null) {
                tmp = p.val;
                p.val = p.next.val;
                p.next.val = tmp;
                p = p.next.next;
            }
            return head;
        }

    }
    

    public static void main(String[] args){
        Solution result = new Solution();
        int [] a = {1, 2, 3, 4, 5};
        ListNode head = new ListNode(a[0]);
        head.buildList(head, a);
        head.printList(head);
        ListNode res = result.swapPairs(head);
        res.printList(res);
    }
}
